### ISSUE TITLE
target_feature cfg only obeys `-C target-feature="+feature"`

### ISSUE THREAD

**Issue by rkruppe** [rust-lang/rust#31662]
[This program](https://gist.github.com/anonymous/f1d2de2d893c74887919) on a 32 bit Linux outputs `no SSE :(` even though our default target *does* enable SSE and SSE2:

```rust
#![feature(cfg_target_feature)]

#[cfg(target_feature="sse")]
fn main() {
    println!("yay, SSE!");
}

#[cfg(not(target_feature="sse"))]
fn main() {
    println!("no SSE :(");
}
```

I'll save myself the hassle of copying `rustc -vV` from the VM I tested this on, because the relevant code hasn't been touched since `cfg_target_feature` was introduced in 4f44258 (July 2015).

This is because the feature detection is really naive, neither asking LLVM for details nor even looking at rustc's own target definitions. @huonw mentioned this in https://github.com/rust-lang/rust/issues/27731#issuecomment-163977414 but IMHO this deserves an issue of its own.

**Comment by huonw** [comment]
Yeah, this should be looking at the target spec in use (i.e. even if it is a custom one, not knowing about the built-in ones), and it should also correctly turn off features when requested.

> neither asking LLVM for details

(I tried to do this, but couldn't find a way to actually get it to work. This would be far-and-away the nicest way.)

**Comment by Aatch** [comment]
Argh, LLVM apparently doesn't want anybody to get/query CPU features. This means it'll require a patch to LLVM to do properly, not a big one, since the functionality we need is basically already there (is feature X enabled), but it's not exposed at present.

**Comment by rkruppe** [comment]
Querying LLVM would be perfect, but since we're committed to supporting some older releases (not sure which ones exactly) this won't be a solution in the short term. Still, if someone wrote that patch and got it upstream *now*, this issue could be solved properly in the future (whenever we bump our minimum requirement to, let's say, 3.9).

**Comment by ranma42** [comment]
I think it might be possible to extract the features from LLVM without any changes to its code base.
Given an `llvm::TargetMachine` they are made available through its `getMCSubtargetInfo()` method. The `MCSubtargetInfo` objects exposes the raw features through the `getFeatureBits ()` method and (given the CPU, which is available from the same object though `getCPU()`), they can be associated to the corresponding strings using `llvm::SubtargetFeatures`.

**Comment by huonw** [comment]
Fixed by #31709, I believe. Thanks @ranma42!
